### PR TITLE
Triplet Loss - third attempt

### PR DESCRIPTION
**Pull request by TuSKan**
Hi Jan, it's me, again :)

The only implementation of triplet loss semi-hard is on _tensorflow-addons_ (deprecated), without port to _Keras_. Every other implementation, (_PyTorch_, ...) is a copy of that one. And the main problem of this implementation it's a creepy code.

When I was thinking to contribute with your package, A loss function could be the easier contribution. In my mind it will be a simple recode from _python_ to _golang_. I had no idea that was need to studding in so deep detail, making exercises on paper to figure out what's happening on code.

Now it's my third PR attempt. I hope it's getting better. Maybe (for sure) there are some (many) mistakes on this version, but now after so many hours, I decide to push and see your opinion on that.

Any comments and suggestions are greatly appreciated.

### CODE REVIEW

**Comment by janpfeifer** [comment]
Thank you for the contribution Rener!

I hear you about the difficulty. I re-implemented a bunch of things in the same way (from TensorFlow, Keras, PyTorch), and every now and then I had to change it a lot. Specially, for old code 😒 

The 2nd PR was already great -- I thought you could use that. And btw, if you don't want to do this adjustment to GoMLX "conventions", let me know. It's good as is, we can just merge and I can do these nitty changes here.

Let me quickly review it!

**Comment by janpfeifer** [comment]
Thank you it looks great 😃 

Also ... I hope that having to go deeper on the math you get a better understanding and learn more about it. This things come in handy later on.

**Comment by janpfeifer** [comment]
Btw, thanks for the fix in `MaskedReduceMin` ... clearly I never used it before :)

**Comment by janpfeifer** [comment]
Thanks I will do it. I notice one of the tests is failing, so I'll check it out :)

Some numeric differences are expected. If the values are really different, then one needs to printout the intermediary results to see what is happening.

In GoMLX it's easy. Any `Node`, just do `node.SetMessage("...")` and it gets logged when executed.
